File: Freedom/微信/通讯录/WXNewFriendViewController.m
//  TLNewFriendViewController.m
//  Freedom
// Created by Super
#import "WXNewFriendViewController.h"
#import "WXSearchController.h"
#import "WXAddFriendViewController.h"
#import "WXContactsViewController.h"
#import "WXTableViewCell.h"
static const NSString *TLThirdPartFriendTypeContacts = @"1";
static const NSString *TLThirdPartFriendTypeQQ = @"2";
static const NSString *TLThirdPartFriendTypeGoogle = @"3";
@interface WXAddThirdPartFriendCell : WXTableViewCell
@property (nonatomic, assign) id<WXAddThirdPartFriendCellDelegate>delegate;
/*第三方类型
 *  {
 *      TLThirdPartFriendTypeContacts
 *      TLThirdPartFriendTypeQQ
 *      TLThirdPartFriendTypeGoogle
 *  }*/
@property (nonatomic, strong) NSArray *thridPartItems;
@end
@interface WXAddThirdPartFriendItem : UIButton
/*第三方类型
 *  {
 *      TLThirdPartFriendTypeContacts
 *      TLThirdPartFriendTypeQQ
 *      TLThirdPartFriendTypeGoogle
 *  }*/
@property (nonatomic, strong) NSString *itemTag;
- (id)initWithImagePath:(NSString *)imagePath andTitle:(NSString *)title;
@end
@interface WXAddThirdPartFriendItem ()
@property (nonatomic, strong) UIImageView *iconImageView;
@property (nonatomic, strong) UILabel *textLabel;
@end
@implementation WXAddThirdPartFriendItem
- (id)initWithImagePath:(NSString *)imagePath andTitle:(NSString *)title{
    if (self = [super initWithFrame:CGRectZero]) {
        [self.iconImageView setImage:[UIImage imageNamed:imagePath]];
        [self.textLabel setText:title];
        [self addSubview:self.iconImageView];
        [self addSubview:self.textLabel];
        [self p_addMasonry];
    }
    return self;
}
#pragma mark - Pirvate Methods -
- (void)p_addMasonry{
    [self.iconImageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.mas_equalTo(self).mas_offset(10);
        make.centerX.mas_equalTo(self);
    }];
    [self.textLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.mas_equalTo(self.iconImageView.mas_bottom).mas_equalTo(5);
        make.centerX.mas_equalTo(self.iconImageView);
    }];
}
#pragma mark - Getter -
- (UIImageView *)iconImageView{
    if (_iconImageView == nil) {
        _iconImageView = [[UIImageView alloc] init];
    }
    return _iconImageView;
}
- (UILabel *)textLabel{
    if (_textLabel == nil) {
        _textLabel = [[UILabel alloc] init];
        [_textLabel setFont:[UIFont systemFontOfSize:12.0f]];
    }
    return _textLabel;
}
@end
@interface WXAddThirdPartFriendCell ()
@property (nonatomic, strong) NSDictionary *itemsDic;
@property (nonatomic, strong) WXAddThirdPartFriendItem *contacts;
@property (nonatomic, strong) WXAddThirdPartFriendItem *qq;
@property (nonatomic, strong) WXAddThirdPartFriendItem *google;
@end
@implementation WXAddThirdPartFriendCell
- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        [self setSelectionStyle:UITableViewCellSelectionStyleNone];
        [self setBottomLineStyle:TLCellLineStyleFill];
        _itemsDic = @{
                      TLThirdPartFriendTypeContacts : self.contacts,
                      TLThirdPartFriendTypeQQ : self.qq,
                      TLThirdPartFriendTypeGoogle : self.google};
    }
    return self;
}
- (void)setThridPartItems:(NSArray *)thridPartItems{
    if (_thridPartItems == thridPartItems) {
        return;
    }
    _thridPartItems = thridPartItems;
    WXAddThirdPartFriendItem *lastItem;
    for(UIView *v in self.contentView.subviews){
        [v removeFromSuperview];
    }
    for (int i = 0; i < thridPartItems.count; i++) {
        NSString *keyStr = [thridPartItems objectAtIndex:i];
        WXAddThirdPartFriendItem *item = [self.itemsDic objectForKey:keyStr];
        [self.contentView addSubview:item];
        [item mas_remakeConstraints:^(MASConstraintMaker *make) {
            make.top.and.bottom.mas_equalTo(self.contentView);
        }];
        if (i == 0) {
            [item mas_updateConstraints:^(MASConstraintMaker *make) {
                make.left.mas_equalTo(self.contentView);
            }];
        }else{
            [item mas_updateConstraints:^(MASConstraintMaker *make) {
                make.left.mas_equalTo(lastItem.mas_right);
                make.width.mas_equalTo(lastItem);
            }];
        }
        if (i == self.thridPartItems.count - 1) {
            [item mas_updateConstraints:^(MASConstraintMaker *make) {
                make.right.mas_equalTo(self.contentView);
            }];
        }
        lastItem = item;
    }
    [self setNeedsDisplay];
}
- (void)drawRect:(CGRect)rect{
    [super drawRect:rect];
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetLineWidth(context, 0.5);
    CGContextSetStrokeColorWithColor(context, [UIColor grayColor].CGColor);
    CGContextBeginPath(context);
    if (self.thridPartItems.count == 2) {
        CGContextMoveToPoint(context, self.frame.size.width / 2.0, 0);
        CGContextAddLineToPoint(context, self.frame.size.width / 2.0, self.frameHeight);
    }else if (self.thridPartItems.count == 3) {
        CGContextMoveToPoint(context, self.frame.size.width / 3.0, 0);
        CGContextAddLineToPoint(context, self.frame.size.width / 3.0, self.frameHeight);
        CGContextMoveToPoint(context, self.frame.size.width / 3.0 * 2, 0);
        CGContextAddLineToPoint(context, self.frame.size.width / 3.0 * 2, self.frameHeight);
    }
    CGContextStrokePath(context);
}
#pragma mark - Event Response -
- (void)itemButtonDown:(WXAddThirdPartFriendItem *)item{
    if (_delegate && [_delegate respondsToSelector:@selector(addThirdPartFriendCellDidSelectedType:)]) {
        [_delegate addThirdPartFriendCellDidSelectedType:item.itemTag];
    }
}
#pragma mark - Getter -
- (WXAddThirdPartFriendItem *)contacts{
    if (_contacts == nil) {
        _contacts = [[WXAddThirdPartFriendItem alloc] initWithImagePath:@"newFriend_contacts" andTitle:@"添加手机联系人"];
        _contacts.itemTag = [TLThirdPartFriendTypeContacts copy];
        [_contacts addTarget:self action:@selector(itemButtonDown:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _contacts;
}
- (WXAddThirdPartFriendItem *)qq{
    if (_qq == nil) {
        _qq = [[WXAddThirdPartFriendItem alloc] initWithImagePath:@"newFriend_qq" andTitle:@"添加QQ好友"];
        _qq.itemTag = [TLThirdPartFriendTypeQQ copy];
        [_qq addTarget:self action:@selector(itemButtonDown:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _qq;
}
- (WXAddThirdPartFriendItem *)google{
    if (_google == nil) {
        _google = [[WXAddThirdPartFriendItem alloc] initWithImagePath:@"newFriend_google" andTitle:@"添加Google好友"];
        _google.itemTag = [TLThirdPartFriendTypeGoogle copy];
        [_google addTarget:self action:@selector(itemButtonDown:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _google;
}
@end
@interface WXNewFriendSearchViewController : WXTableViewController <UISearchResultsUpdating>
@end
@implementation WXNewFriendSearchViewController
#pragma mark - Delegate -
//MARK: UISearchResultsUpdating
- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
    
}
@end
@interface WXNewFriendViewController ()
@property (nonatomic, strong) WXSearchController *searchController;
@property (nonatomic, strong) WXNewFriendSearchViewController *searchVC;
@end
@implementation WXNewFriendViewController
- (void)viewDidLoad{
    [super viewDidLoad];
    [self.navigationItem setTitle:@"新的朋友"];
    [self.view setBackgroundColor:[UIColor whiteColor]];
    [self.tableView setTableHeaderView:self.searchController.searchBar];
    
    UIBarButtonItem *rightBarButton = [[UIBarButtonItem alloc] initWithTitle:@"添加朋友" style:UIBarButtonItemStylePlain target:self action:@selector(rightBarButtonDown:)];
    [self.navigationItem setRightBarButtonItem:rightBarButton];
    
    [self registerCellClass];
}
#pragma mark - Event Response -
- (void)rightBarButtonDown:(UIBarButtonItem *)sender{
    WXAddFriendViewController *addFriendVC = [[WXAddFriendViewController alloc] init];
    [self setHidesBottomBarWhenPushed:YES];
    [self.navigationController pushViewController:addFriendVC animated:YES];
}
#pragma mark - Getter -
- (WXSearchController *)searchController{
    if (_searchController == nil) {
        _searchController = [[WXSearchController alloc] initWithSearchResultsController:self.searchVC];
        [_searchController setSearchResultsUpdater:self.searchVC];
        [_searchController.searchBar setPlaceholder:@"微信号/手机号"];
        [_searchController.searchBar setDelegate:self];
    }
    return _searchController;
}
- (WXNewFriendSearchViewController *)searchVC{
    if (_searchVC == nil) {
        _searchVC = [[WXNewFriendSearchViewController alloc] init];
    }
    return _searchVC;
}
#pragma mark - Private Methods -
- (void)registerCellClass{
    [self.tableView registerClass:[WXAddThirdPartFriendCell class] forCellReuseIdentifier:@"TLAddThirdPartFriendCell"];
}
#pragma mark - Delegate -
//MARK: UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 2;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    if (section == 0) {
        return 1;
    }
    return 0;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == 0) {
        WXAddThirdPartFriendCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TLAddThirdPartFriendCell"];
        [cell setThridPartItems:@[TLThirdPartFriendTypeContacts]];
        [cell setDelegate:self];
        return cell;
    }
    return nil;
}
//MARK: UITableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == 0) {
        return 80;
    }
    return 60.0f;
}
- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section{
    if (section == 0) {
        return 20.0f;
    }
    return 0.0f;
}
//MARK: UISearchBarDelegate
- (void)searchBarTextDidBeginEditing:(UISearchBar *)searchBar{
    
}
- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText{
}
- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar{
}
//MARK: TLAddThirdPartFriendCellDelegate
- (void)addThirdPartFriendCellDidSelectedType:(NSString *)thirdPartFriendType{
    if ([TLThirdPartFriendTypeContacts isEqualToString:thirdPartFriendType]) {
        WXContactsViewController *contactsVC = [[WXContactsViewController alloc] init];
        [self setHidesBottomBarWhenPushed:YES];
        [self.navigationController pushViewController:contactsVC animated:YES];
    }else if ([TLThirdPartFriendTypeQQ isEqualToString:thirdPartFriendType]) {
        
    }else if ([TLThirdPartFriendTypeGoogle isEqualToString:thirdPartFriendType]) {
        
    }
}
@end

File: Freedom/微信/通讯录/WXNewFriendViewController.swift
//
//  WXNewFriendViewController.swift
//  Freedom

import Foundation
protocol WXAddThirdPartFriendCellDelegate: NSObjectProtocol {
    func addThirdPartFriendCellDidSelectedType(_ type: TLThirdPartType)
}
class WXAddThirdPartFriendCell: WXTableViewCell {
    weak var delegate: WXAddThirdPartFriendCellDelegate?
    var lastItem: WXAddThirdPartFriendItem?
    var thridPartItems: [WXAddThirdPartFriendItem] = [] {
        didSet {
            for v: UIView in contentView.subviews {
                v.removeFromSuperview()
            }
            for i in 0..<thridPartItems.count {
                let keyStr = thridPartItems[i].itemType
                let item = itemsDic[keyStr.rawValue]
                if let anItem = item {
                    contentView.addSubview(anItem)
                }
                item?.snp.remakeConstraints({ (make) in
                    make.top.bottom.equalTo(self.contentView)
                })
                if i==0{
                    item?.snp.updateConstraints({ (make) in
                        make.left.equalTo(self.contentView)
                    })
                }else{
                    item?.snp.updateConstraints({ (make) in
                        make.left.equalTo(lastItem?.snp.right ?? 0)
                        make.width.equalTo(lastItem ?? 0)
                    })
                }
                if i == self.thridPartItems.count - 1 {
                    item?.snp.updateConstraints({ (make) in
                        make.right.equalTo(self.contentView)
                    })
                }
                lastItem = item
            }
            setNeedsDisplay()
        }
    }
    private var itemsDic: [String : WXAddThirdPartFriendItem] = [:]
    private lazy var contacts: WXAddThirdPartFriendItem = {
        let contacts = WXAddThirdPartFriendItem(imagePath: "newFriend_contacts", andTitle: "添加手机联系人")
        contacts.itemType = .contacts
        contacts.addTarget(self, action: #selector(self.itemButtonDown(_:)), for: .touchUpInside)
        return contacts
    }()
    private lazy var qq: WXAddThirdPartFriendItem = {
        let qq = WXAddThirdPartFriendItem(imagePath: "newFriend_qq", andTitle: "添加QQ好友")
        qq.itemType = .QQ
        qq.addTarget(self, action: #selector(self.itemButtonDown(_:)), for: .touchUpInside)
        return qq
    }()
    private lazy var google: WXAddThirdPartFriendItem = {
        let google = WXAddThirdPartFriendItem(imagePath: "newFriend_google", andTitle: "添加Google好友")
        google.itemType = .google
        google.addTarget(self, action: #selector(self.itemButtonDown(_:)), for: .touchUpInside)
        return google
    }()

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        selectionStyle = .none
        bottomLineStyle = .fill
        itemsDic = [TLThirdPartType.contacts.rawValue: contacts, TLThirdPartType.QQ.rawValue: qq, TLThirdPartType.google.rawValue: google]
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    override func draw(_ rect: CGRect) {
        super.draw(rect)
        guard let context = UIGraphicsGetCurrentContext() else { return }
        context.setLineWidth(0.5)
        context.setStrokeColor(UIColor.gray.cgColor)
        context.beginPath()
        if thridPartItems.count == 2 {
            context.move(to: CGPoint(x: frame.size.width / 2.0, y: 0))
            context.addLine(to: CGPoint(x: frame.size.width / 2.0, y: frame.size.height))
        } else if thridPartItems.count == 3 {
            context.move(to: CGPoint(x: frame.size.width / 3.0, y: 0))
            context.addLine(to: CGPoint(x: frame.size.width / 3.0, y: frame.size.height))
            context.move(to: CGPoint(x: frame.size.width / 3.0 * 2, y: 0))
            context.addLine(to: CGPoint(x: frame.size.width / 3.0 * 2, y: frame.size.height))
        }
        context.strokePath()
    }
    @objc func itemButtonDown(_ item: WXAddThirdPartFriendItem) {
        delegate?.addThirdPartFriendCellDidSelectedType(item.itemType)
    }
}
class WXNewFriendSearchViewController: WXTableViewController, UISearchResultsUpdating {
    func updateSearchResults(for searchController: UISearchController) {
    }
}
class WXAddThirdPartFriendItem: UIButton {
    var itemType = TLThirdPartType.contacts
    private var iconImageView = UIImageView()
    private var textLabel = UILabel()
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    init(imagePath: String, andTitle title: String) {
        super.init(frame: CGRect.zero)
        textLabel.font = UIFont.systemFont(ofSize: 12.0)
        iconImageView.image = UIImage(named: imagePath)
        textLabel.text = title
        addSubview(iconImageView)
        addSubview(textLabel)
        iconImageView.snp.makeConstraints { (make) in
            make.top.equalToSuperview().offset(10)
            make.centerX.equalToSuperview()
        }
        textLabel.snp.makeConstraints { (make) in
            make.top.equalTo(self.iconImageView.snp.bottom).offset(5)
            make.centerX.equalTo(self.iconImageView)
        }
    }
}
class WXNewFriendViewController: WXTableViewController, UISearchBarDelegate, WXAddThirdPartFriendCellDelegate {
    private var searchVC = WXNewFriendSearchViewController()
    private lazy var searchController: WXSearchController = {
        let searchController = WXSearchController(searchResultsController: searchVC)
        searchController.searchResultsUpdater = searchVC
        searchController.searchBar.placeholder = "微信号/手机号"
        searchController.searchBar.delegate = self
        return searchController
    }()
    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "新的朋友"
        view.backgroundColor = UIColor.white
        tableView.tableHeaderView = searchController.searchBar
        let rightBarButton = UIBarButtonItem(title: "添加朋友", style: .plain, target: self, action: #selector(WXNewFriendViewController.rightBarButtonDown(_:)))
        navigationItem.rightBarButtonItem = rightBarButton
        tableView.register(WXAddThirdPartFriendCell.self, forCellReuseIdentifier: "TLAddThirdPartFriendCell")
    }
    @objc func rightBarButtonDown(_ sender: UIBarButtonItem) {
        let addFriendVC = WXAddFriendViewController()
        hidesBottomBarWhenPushed = true
        navigationController?.pushViewController(addFriendVC, animated: true)
    }
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            return 1
        }
        return 0
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 0 {
            let cell = tableView.dequeueReusableCell(withIdentifier: "TLAddThirdPartFriendCell") as! WXAddThirdPartFriendCell
            let item = WXAddThirdPartFriendItem(imagePath: "", andTitle: "")
            item.itemType = TLThirdPartType.contacts
            cell.thridPartItems = [item]
            cell.delegate = self
            return cell
        }
        return UITableViewCell()
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return indexPath.section == 0 ? 80 : 60.0
    }
    override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return section == 0 ? 20 : 0.0
    }
    func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
    }
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
    }
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
    }
    func addThirdPartFriendCellDidSelectedType(_ type: TLThirdPartType) {
        switch type {
        case .contacts:
            let contactsVC = WXContactsViewController()
            hidesBottomBarWhenPushed = true
            navigationController?.pushViewController(contactsVC, animated: true)
        case .QQ:break
        case .google:break
        }
    }
}
